File: sources/unity/Assets/Plugins/General.Typescript/core/namespace/Namespace.cs
using System;

namespace General.Typescript
{
    public class Namespace : Base
    {
        internal Namespace(IntPtr context, IntPtr handle, string name, Base parent) : base(context, handle, name, parent)
        {
            this.registerName();
        }

        public Namespace DeclareNamespace(string name)
        {
            return new Namespace(mContext, Entry.Namespace.DeclareNamespace(this, name), name, this);
        }

        public Class<T> DeclareClass<T>()
        {
            return Class.Create<T>(mContext, Entry.Namespace.DeclareClass(this, typeof(T)), this);
        }

        public Enum<T> DeclareEnum<T>()
        {
            return new Enum<T>(mContext, Entry.Namespace.DeclareEnum(this, typeof(T).Name), this);
        }
    }
}

File: sources/platforms/typescript/core/namespace/Namespace.cpp
#include "Namespace.h"
#include "../../extra/log/Log.h"
#include "../class/Class.h"
#include "../enum/Enum.h"

ASSIGN_BASE_CHILD_TYPE(TypescriptNamespace, BaseChildTypes::BaseChildType_Namespace);

TypescriptNamespace* TypescriptNamespace::Create(const std::string& name, Base* parent)
{
	return Base::Create<TypescriptNamespace>(name, parent);
}

TypescriptNamespace::TypescriptNamespace(const std::string& name, Base* parent) : Base(TYPE, name, parent) { }

TypescriptNamespace::~TypescriptNamespace() { }

TypescriptNamespace* TypescriptNamespace::DeclareNamespace(const std::string& name)
{
	TypescriptNamespace* instance = nullptr;
	if (mNamespaces.end() == mNamespaces.find(name))
	{
		instance = TypescriptNamespace::Create(name, this);
		mNamespaces[name] = instance;
		instance->Bind();
	}
	else
	{
		LogError("There is already an namespace called %s!", name.c_str());
	}
	return instance;
}

TypescriptClass* TypescriptNamespace::DeclareClass(const std::string& name, const std::string& base, const bool& hasConstructor, const bool& isStatic)
{
	TypescriptClass* instance = nullptr;
	if (mClasses.end() == mClasses.find(name))
	{
		instance = TypescriptClass::Create(name, base, hasConstructor, isStatic, this);
		mClasses[name] = instance;
        instance->Bind();
	}
	else
	{
		LogError("There is already a class called %s!", name.c_str());
	}
	return instance;
}

TypescriptEnum* TypescriptNamespace::DeclareEnum(const std::string& name)
{
	TypescriptEnum* instance = nullptr;
	if (mEnums.end() == mEnums.find(name))
	{
		instance = TypescriptEnum::Create(name, this);
		mEnums[name] = instance;
		instance->Bind();
	}
	else
	{
		LogError("There is already a enum called %s!", name.c_str());
	}
	return instance;
}
